WIP: MWB-194 Tox integration

CaptainOfHacks opened the pull request.

schivmeister committed:
Fix failing prefix lookup test

Strangely, the reverse lookup for http://www.w3.org/2004/02/skos/core#
does not work. It returns only dc. This can be reproduced via the
browser as well, which redirects to a page of alternative results
instead of SKOS.

Try: http://prefix.cc/reverse?uri=http://xmlns.com/foaf/0.1/&format=json

A JSON result is returned as expected with expected items.

Then try:
http://prefix.cc/reverse?uri=http://www.w3.org/2004/02/skos/core%23&format=json

A JSON result is returned as expected, but it contains only the
(unexpected) dc prefix and corresponding namespace.

Go on to http://prefix.cc/ and enter
http://www.w3.org/2004/02/skos/core# in the search bar. The redirected
page shows dc as the canonical and lists skos as the last alternative
(as of date).

Somebody should report this but for now, fall back to using the DC pair
that's being returned, as the test.

schivmeister committed:
Initial Tox configuration, Python 3.8

All tests pass for Python 3.8, against present list of requirements,
as-is.

However, for some reason we do not yet know, Tox cannot run `setup.py`
or the setup can't locate the requirements files in the Tox environment.
We therefore use `skipsdist=true` for now, leaving the packaging
untested.

See https://github.com/gitpython-developers/GitPython/issues/1090

CaptainOfHacks committed:
add tox.ini

CaptainOfHacks committed:
Update requirements-dev.txt

CaptainOfHacks committed:
update github CI/CD workflows

CaptainOfHacks committed:
Update Makefile

schivmeister committed:
Fix yet another failing test - status code checker

Not sure when was the last time https://httpstat.us/ worked, but it
doesn't seem to be up anymore or it is unstable. Use
https://httpbin.org/ instead.

schivmeister committed:
Fix another failing test for Python 3.10+

Python 3.10 onwards dropped support for importing `Mapping` from
`collections` directly. Instead, one needs to import from
`collections.abc`. This has been supported since Python 3.3, so is
backward-compatible with the last LTS Python 3.8.

schivmeister commented:
please ignore this PR for the time being, it is being used to collaborate for tox-related changes with @CaptainOfHacks, checking the CI/CD pipeline compatibility. also contains unrelated changes, which we may move over to another PR.

schivmeister committed:
Support latest compatible Pandas for Python 3.8

Pandas 2.0.3 is the last version that supports Python version range 3.8,
3.11. Starting from 1.5.3, specifying `value` is not supported for
`df.replace()` when using a nested _dict_ as the `to_replace` value. In
such cases, missing value fillers (e.g. using `np.nan`) have to be
included in the dictionary itself.

Also start ignoring some test output artifacts.

schivmeister committed:
Some CI changes for multiple Py envs, code clarity

- Run GitHub CI for only used/expected branches in project
- Add/modify some comments in Github CI config for clarity
- Add GitHub CI matrix strategy for multiple Python environments
- Update Actions versions and stick to specific versions
- Add a make install target for avoiding redundant pip installs
- Use a recent Python version for publishing workflow

Note that there may be some redundancy between GitHub's matrix strategy
support for multiple environments and Tox. Combining both usually
accommodates sophisticated configurations (like multiple Django
sub-environments for a particular Python environment), which we don't
yet have a need for.

schivmeister committed:
Soft-lock dependency versions

Lock to the latest packages of the lowest-supported Python (3.8).
Explicit dependencies only, so as to not confuse people (which is what
`pip freeze` is often good at doing). Use a trick like below to
manually/ChatGPT inspect/replace existing requirements:

`grep -wiFf $requirements_file <(pip freeze)`

Remove version specifiers in the requirement files and upgrade
dependencies first, then compare the two lists. Pip dependencies like
`setuptools` and `wheel` can be and are omitted. We might want to move
to listing only parent packages, or to Poetry :)

schivmeister committed:
Support CI build on Mac and Windows

Attain true cross-platform compatibility by making Github Actions CI run
for all three major operating systems, for all runtimes.

- Add multi-dimensional environments in CI config
- Add job concurrency group (or risk being killed)

**Caveat**: Tox appears to run twice on Windows, and sometimes on Mac.
This may be because of the system-wide availability of multiple
interpreters in some runner types. We either need to decouple Tox from
the build (take it out of Makefile) and use the Tox GH Actions plug-in,
or add some condition to the job to run Tox for only one
runtime/interpreter version.

schivmeister committed:
Reintroduce common build dependencies

While `setuptools` and `wheel` should be installed by default in recent
Python environments, make no assumption as there are various GNU/Linux
distributions, for example, that may not have the same packaging
practices.

schivmeister committed:
New version 0.2.0, dropping support for Python <= 3.7

Python 3.7 and earlier reached EOL earlier this year, and many
tools/libraries have stopped supporting those runtime. Continuing to
support an old runtime will be near impossible due to dependency hell.
It is therefore only wise to follow suit.

- Update code, tests
- Update CI config
- Update dependencies
- Update metadata

We will aim to support and test against the oldest `security` release
and the latest `bugfix` release, less one (if it's too new and not a lot
of tools have adopted it).

The jump from Python 3.6 to 3.8 doesn't involve that many changes among
dependency APIs for this project -- there is just a
`collections.Mapping` namespace change and a Pandas
`DataFrame.replace()` field behaviour change.

Users are encouraged to upgrade even if there is no API or ABI change in
this project itself, as an EOL runtime is a security risk.

schivmeister commented:
A new PR will be created while this is kept for historical record (bug filed with GitHub for pipeline anomaly).
